File: backend/src/main/java/com/team05/petmeeting/domain/ads/client/InstagramClient.java
package com.team05.petmeeting.domain.ads.client;

import com.team05.petmeeting.domain.ads.config.InstagramProperties;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientResponseException;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

@Component
@RequiredArgsConstructor
public class InstagramClient {

    private final InstagramProperties properties;

    private static final String BASE_URL = "https://graph.facebook.com/v19.0";

    // 1단계: 미디어 컨테이너 생성
    public String createMediaContainer(String imageUrl, String caption) {
        String url = BASE_URL + "/" + properties.getUserId() + "/media"
                + "?image_url=" + imageUrl  // URLEncoder 제거
                + "&caption=" + caption
                + "&access_token=" + properties.getAccessToken();

        try {
            return RestClient.create()
                    .post()
                    .uri(url)
                    .retrieve()
                    .body(String.class);
        } catch (RestClientResponseException e) {
            throw new IllegalStateException(
                    "미디어 컨테이너 생성 실패: %s".formatted(e.getResponseBodyAsString()), e
            );
        }
    }

    // 2단계: 미디어 게시
    public String publishMedia(String containerId) {
        String url = BASE_URL + "/" + properties.getUserId() + "/media_publish";

        String body = "creation_id=" + containerId
                + "&access_token=" + properties.getAccessToken();

        try {
            return RestClient.create()
                    .post()
                    .uri(url)
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .body(body)
                    .retrieve()
                    .body(String.class);
        } catch (RestClientResponseException e) {
            throw new IllegalStateException(
                    "미디어 게시 실패: %s".formatted(e.getResponseBodyAsString()), e
            );
        }
    }
}

File: backend/src/main/java/com/team05/petmeeting/domain/ads/client/InstagramClient.kt
package com.team05.petmeeting.domain.ads.client

import com.team05.petmeeting.domain.ads.config.InstagramProperties
import org.springframework.stereotype.Component
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestClientResponseException
import org.springframework.web.util.UriComponentsBuilder

@Component
class InstagramClient(
    private val properties: InstagramProperties
) {
    // 1단계: 미디어 컨테이너 생성
    fun createMediaContainer(imageUrl: String?, caption: String?): String? {
        val userId = requireInstagramProperty(properties.userId, "Instagram User ID")
        val accessToken = requireInstagramProperty(properties.accessToken, "Instagram Access Token")
        val mediaImageUrl = imageUrl?.takeIf { it.isNotBlank() }
            ?: throw IllegalArgumentException("인스타그램 이미지 URL이 비어있습니다.")
        val url = UriComponentsBuilder
            .fromUriString("$BASE_URL/$userId/media")
            .queryParam("image_url", mediaImageUrl)
            .queryParam("caption", caption.orEmpty())
            .queryParam("access_token", accessToken)
            .build()
            .toUriString()

        try {
            return RestClient.create()
                .post()
                .uri(url)
                .retrieve()
                .body(String::class.java)
        } catch (e: RestClientResponseException) {
            throw IllegalStateException(
                "미디어 컨테이너 생성 실패: ${e.responseBodyAsString}", e
            )
        }
    }

    // 2단계: 미디어 게시
    fun publishMedia(containerId: String): String? {
        val userId = requireInstagramProperty(properties.userId, "Instagram User ID")
        val accessToken = requireInstagramProperty(properties.accessToken, "Instagram Access Token")
        val url = "$BASE_URL/$userId/media_publish"

        val body = "creation_id=$containerId&access_token=$accessToken"

        try {
            return RestClient.create()
                .post()
                .uri(url)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .body(body)
                .retrieve()
                .body(String::class.java)
        } catch (e: RestClientResponseException) {
            throw IllegalStateException(
                "미디어 게시 실패: ${e.responseBodyAsString}", e
            )
        }
    }

    private fun requireInstagramProperty(value: String?, name: String): String {
        return value?.takeIf { it.isNotBlank() }
            ?: throw IllegalStateException("${name}가 설정되지 않았습니다.")
    }

    companion object {
        private const val BASE_URL = "https://graph.facebook.com/v19.0"
    }
}
